src/onepass.rs

```diff
@@ -0,0 +1,1187 @@
+            table_size += num_byte_classes;
+            match inst {
+                &Inst::EmptyLook(_) | &Inst::Save(_) =>
+                    table_size += num_byte_classes,
+                _ => {}
+            }
+            if table_size > max_table_size {
+                return Err(OnePassError::TooBig);
+            }
+        }
+
+        trace!("new compiler for:\n{:?}", prog);
+        Ok(OnePassCompiler {
+            table: vec![],
+            num_byte_classes: num_byte_classes,
+            only_utf8: only_utf8,
+
+            transitions: {
+                let mut x = Vec::new();
+                for _ in 0..prog.len() {
+                    x.push(None);
+                }
+                x
+            },
+            accepting_states: vec![false; prog.len()],
+            prog: prog,
+        })
+    }
+
+    /// Attempt to compile the regex to a OnePass DFA
+    pub fn compile(mut self) -> Result<OnePass, OnePassError> {
+        // A DAG of forwarding relationships indicating when
+        // a state needs to be forwarded to an Action state
+        // once that Action state has been fully constructed.
+        let mut forwards = Forwards::new();
+
+        // Compute the prioritized transition tables for all of the
+        // instructions which get states.
+        let mut state_edge = vec![0];
+        while let Some(i) = state_edge.pop() {
+            state_edge.extend(self.inst_trans(i, &mut forwards)?);
+        }
+
+        // Solve the dependency relationships between all the
+        // forwarding directives that were emitted by inst_trans.
+        for fwd in forwards.into_iter_topo() {
+            self.perform_forward(fwd?);
+        }
+
+        // Now emit the transitions in a form that we can actually
+        // execute.
+        self.bake_transitions();
+
+        Ok(OnePass {
+            table: self.table,
+            prefixes: self.prog.prefixes,
+            num_byte_classes: self.num_byte_classes,
+            byte_classes: self.prog.byte_classes,
+            start_state: 0 | STATE_ACTION,
+            is_anchored_start: self.prog.is_anchored_start,
+            is_anchored_end: self.prog.is_anchored_end,
+            only_utf8: self.only_utf8,
+        })
+    }
+
+    /// Compile the stage 1 transition table for the state corresponding
+    /// to the given instruction.
+    ///
+    /// The result of `inst_trans` will end up in `self.transitions`.
+    ///
+    /// Returns a list of child instructions which must be compiled
+    /// via `inst_trans`.
+    fn inst_trans(
+        &mut self,
+        inst_idx: usize,
+        forwards: &mut Forwards,
+    ) -> Result<Vec<usize>, OnePassError> {
+        trace!("::inst_trans inst_idx={}", inst_idx);
+
+        if self.transitions[inst_idx].is_some() {
+            return Ok(vec![]);
+        }
+
+        // Iterate over the children, visiting lower priority
+        // children first.
+        let mut resume = match &self.prog[inst_idx] {
```

Comment:
If the cost of allocating this resume queue for each Bytes instruction is a problem, this might be a candidate for a SmallVec. Probably not worth the dependency without some measurements.